feat(#305): support select one from map

latin-panda opened the pull request:
Closes #305
Closes #258

Related PRs: 
- [Documentation](https://github.com/getodk/docs/pull/1994)
- [Test coverage](https://github.com/getodk/web-forms/pull/509)


### I have verified this PR works in these browsers (latest versions):

- [x] Chrome
- [x] Firefox
- [x] Safari (macOS)
- [ ] Safari (iOS)
- [x] Chrome for Android
- [ ] Not applicable

### What else has been done to verify that this works as intended?

- Manual testing with forms [provided by QA](https://github.com/getodk/web-forms/issues/305#issuecomment-3213346804), and this will go through QA before merging. 
- All question types form works fine.
- New Scenario tests are enabled and passing.
- The automated test coverage [is in this PR](https://github.com/getodk/web-forms/pull/509)

**Test evidence**

<details><summary>Edit submission in Central</summary>

- UI looks good in Central
- Submission renders correctly
- Edit submission works

https://github.com/user-attachments/assets/c4d622f3-90f8-4ef3-b88e-0a347a653e4c

</details>

<details><summary>Empty external files </summary>

- Renders empty map

<img width="974" height="594" alt="empty-csv-file" src="https://github.com/user-attachments/assets/0e595a61-9260-4d68-a444-791600db0d34" />
<img width="982" height="834" alt="empty-geojson-file" src="https://github.com/user-attachments/assets/faa5b824-142d-48b5-bee5-f0a1d445cda5" />

</details>

<details><summary>Missing external files </summary>

- Shows error modal - standard for all missing attachments

<img width="700" height="394" src="https://github.com/user-attachments/assets/fc126ddf-66d3-4d9a-8d20-663f5d549697" />

</details>

<details><summary>Map dynamic bundle</summary>


Loads map bundle when form has a map question

<img width="800" height="921" alt="no-bundle-form-without-map" src="https://github.com/user-attachments/assets/ac51f519-31a5-4c9d-8c92-3b0d1fb4524b" />

It doesn't load map bundle when form doesn't have map question 

<img width="800" height="924" alt="loads-bundle-form-with-map" src="https://github.com/user-attachments/assets/9f31831e-a08a-417b-8951-76276eadf038" />

Show error when it cannot load bundle

https://github.com/user-attachments/assets/83048e87-c1f1-4491-825b-a9c8d764cf4e

</details>

<details><summary>Map control buttons</summary>

https://github.com/user-attachments/assets/e17863af-b742-4e23-99a7-01767a602416

</details>

<details><summary>Renders special characters</summary>

<img width="800" height="601" alt="Screenshot 2025-10-02 at 6 06 38 PM" src="https://github.com/user-attachments/assets/c2d3ecf0-8ad6-4367-891a-c9bfd406e3f4" />

</details>


### Why is this the best possible solution? Were any other approaches considered?


### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
- Preexisting select features are unaffected (ItemSet, Items selects, select with images, select with search, etc)
- The choice filter works fine, and other related functions/calculations.
- Exposes a new select feature 

### Do we need any specific form for testing your changes? If so, please attach one.
[Test plan](https://github.com/getodk/web-forms/issues/305#issuecomment-3206808806)

### What's changed
- Engine support: Exposes geometry and other properties as one array of [string, string] to preserve order when displayed. 
- Rank and select share the same interface for their options, since they use the same XML element.
- The support is only for ItemSets; regular Items don't have additional properties.
- Front-end support: Using OpenLayers, loading async since it's a big bundle. 
- Front-end handles the options and builds a structure that works with OpenLayers. 
- It's structured with smaller components for possible reusability with geo shape / trace in the future.

latin-panda commented on packages/web-forms/src/components/common/map/useMapBlock.ts:
```diff
@@ -0,0 +1,258 @@
+		{ immediate: true }
+	);
+
+	return {
```
@matthew-white this is the core of map and OL. Additional features might be added as a separate module and then imported there. For example:
```
return {
        initializeMap,
		loadGeometries,
		errorMessage,

        ...editShapesModule, // <-- for example something like this.

       ... etc
}
```

garethbowen commented on packages/xforms-engine/src/lib/reactivity/createItemCollection.ts:
```diff
@@ -122,9 +124,20 @@ const createItemsetItems = (
 					});
 					const label = createItemsetItemLabel(context, itemset, value);
 
+					const nodeElements = itemNode
```
> I find it odd to import SelectNode and RankNode in this file to define the Item type on line 23. A cleaner approach might be to define the Item structure here and import it into SelectNode and RankNode. This would avoid dependency inversion and make the structure more intuitive.

Yes I mostly agree. However because SelectItem is exported I think it would be better to define the item interface in `/src/client` somewhere so it is also exported and then extend it for SelectItem and RankItem. This will hopefully make it clear for anyone else using the engine lib.

> This implementation minimize code changes while keeping things functional. The front-end can simply ignore RankNode’s properties. What do you think of this approach? Any suggestions for improvement?

I agree that this type simplification does make life easier here, it is stepping away from the current status quo where the type would enforce that implementers don't access the `properties` of a RankItem. Obviously this will force you to handle it here but hopefully it's not too complicated - just a quick check of the item type before populating properties?

I think we should discuss at some point how much we want types to enforce this sort of checking for other implementers who may want to pick up the engine lib, but in the meantime I'm err on the side of maintaining consistency with the level of checking in interface today. What do you think?

matthew-white reviewed (commented):
I've been working on adapting `useMapBlock()` here for use in Central. It's been really helpful to be able to draw from the code in this PR. Thank you for all the help, @latin-panda!

I've now pushed my PR for Central: getodk/central-frontend#1351. I wanted to leave a few comments about questions I had and a few things I did differently. Overall, I found the code in `useMapBlock()` to be really clear. :+1:

One thing that @latin-panda and I have discussed is that the design/criteria for Web Forms and Central is a bit different. For example:

- Web Forms lets you save features, but Central only lets you select features.
- Central shows the number of features in view (e.g., "Showing 10 of 100").

Web Forms has more functionality than Central, so there's a number of things in this PR that Central doesn't need to do. One thing that I did add to my PR was clustering, since that's important for Central.

I may leave a few more comments later on as code review for my PR progresses, but these are the main things I was thinking about.

Thanks again!

latin-panda commented on packages/web-forms/src/components/common/map/useMapBlock.ts:
```diff
@@ -0,0 +1,258 @@
+	getSelectedStyles,
+	getUnselectedStyles,
+} from '@/components/common/map/map-styles.ts';
+import { Map, View } from 'ol';
```
Yes, they refer to the same class/interface. OL has "index" files in their module, for example this is `ol`: 

<img width="690" height="556" alt="Screenshot 2025-09-29 at 8 12 14 PM" src="https://github.com/user-attachments/assets/77677716-f4d4-4063-81e4-02566c3fa57a" />

I prefer importing from what their "index" exposes. At the end of the day, it's the same.

garethbowen commented on packages/web-forms/src/components/common/map/useMapBlock.ts:
```diff
@@ -0,0 +1,282 @@
+	const selectFeature = (feature: Feature<GeometryType> | undefined): void => {
+		selectedFeature.value = feature;
+
+		featuresVectorLayer.updateStyleVariables({
+			[SELECTED_ID_PROPERTY]: (selectedFeature.value?.get(FEATURE_ID_PROPERTY) as string) ?? '',
+		});
+
+		if (selectedFeature.value != null) {
+			centerFeatureLocation(selectedFeature.value);
+		}
```
I _think_ this would be cleaner if it was updated reactively based on the `selectedFeature.value`, and ditto for the next 3 functions.

Then the `selectFeature` function would just be one line and could be inlined where it's used, right?

garethbowen commented on packages/web-forms/src/components/common/map/useMapBlock.ts:
```diff
@@ -0,0 +1,282 @@
+	const handleClick = (event: MapBrowserEvent) => selectFeatureByPosition(event.pixel);
+
+	const toggleClickBinding = (bindClick: boolean) => {
+		mapInstance?.un('singleclick', handleClick);
```
I think @matthew-white commented somewhere but adding here for visibility... This feels sluggish and using `click` or similar might be better?

matthew-white commented on packages/web-forms/src/components/common/map/useMapBlock.ts:
```diff
@@ -0,0 +1,282 @@
+const GEOLOCATION_TIMEOUT_MS = 10 * 1000;
+const ANIMATION_TIME = 1000;
+const SMALL_DEVICE_WIDTH = 576;
+const MAP_HIT_TOLERANCE = 5;
```
I always manage to select points and shapes when I click on them, but somehow I often miss lines. I click a little too far away, and the selection doesn't go through. Might just be something I'm doing wrong. 😅 But what do you think about increasing the hit tolerance even if just by a pixel or two?

I know you've been updating feature styles as well. Do you think those changes could affect this? I could wait to try it out with the new styles and see if it's still happening for me. Or we could see if the QA team mentions anything.

latin-panda replied:
This hit tolerance wasn't effective. The click only responds to "visible" elements. I've added a style for line features that expands the hit area by including an almost transparent stroke on top of the blue stroke. I've just pushed the change.

matthew-white commented on packages/web-forms/src/components/common/map/useMapBlock.ts:
Just an idea! I just added the following to Central. When you move your mouse over a feature, the cursor changes. I feel like it's a nice hint that the feature is selectable. Here the code would be something like:

```js
mapInstance.on('pointermove', (event) => {
  if (event.dragging) return;
  const hit = mapInstance.hasFeatureAtPixel(event.pixel, {
    hitTolerance: MAP_HIT_TOLERANCE,
    layerFilter: (layer) => layer instanceof WebGLVectorLayer,
  });
  mapInstance.getTargetElement().style.cursor = hit ? 'pointer' : '';
});
```

matthew-white commented on packages/web-forms/src/components/common/map/useMapBlock.ts:
```diff
@@ -0,0 +1,282 @@
+	});
+
+	const featuresSource = new VectorSource();
+	const featuresVectorLayer = new WebGLVectorLayer({
```
For some reason, I had trouble using `WebGLVectorLayer` in testing. Something about the way Central testing is set up, I think. Central is still using Karma + headless Chrome. (We want to move to Vitest — hopefully soon!) Anyway, what I'm doing is detecting whether WebGL2 is supported. If it is, I use `WebGLVectorLayer`, if not, I fall back to `VectorLayer`. I don't think anything needs to change here, since it was just an issue in testing. I just thought I'd mention it, since it was a surprise to me that WebGL might not be supported.

latin-panda replied:
Interesting, WebGL has been around for a few years (approx Chrome v56, Firefox v51). It uses the GPU, so when running tests in headless mode, some WebGL features might not be available, or the test runner might have the GPU disabled (in a Karma configuration). Maybe you can have a look at the karma config. 
I haven't gotten to the automated tests yet. I'll let you know how it goes.

matthew-white replied:
Given that WebGL is widely supported, I bet the issue in Central could be solved by changing the Karma config, like you said. I'm trying to limit the changes I make to Karma, just because we're about to switch it out for Vitest anyway. But I think I'll simplify Central so that instead of feature detection, it's just simple logic like, "if test, use `VectorLayer`, otherwise use `WebGLVectorLayer`."

matthew-white reviewed (commented):
Just a few random thoughts, mostly based on recent changes I've made in Central.

latin-panda commented on packages/web-forms/src/components/common/map/AsyncMap.vue:
```diff
@@ -0,0 +1,234 @@
+	const features: Feature[] = [];
+
+	props.features?.forEach((option) => {
+		try {
```
I've removed the `try/catch` and replaced the `throw`s with warning logs.

latin-panda commented on packages/web-forms/src/components/common/map/AsyncMap.vue:
```diff
@@ -0,0 +1,234 @@
+
+	const first = coords[0];
+	const last = coords[coords.length - 1];
+	return first[0] === last[0] && first[1] === last[1] ? 'Polygon' : 'LineString';
```
Agreed; however, OpenLayers requires an array for coordinates. You reminded me of a type at the top of the file that I forgot to use elsewhere. I've also properly named these `first[0]` chunks.

latin-panda commented on packages/web-forms/src/components/common/map/MapProperties.vue:
```diff
@@ -0,0 +1,127 @@
+
+		<div class="map-properties-content">
+			<div v-for="[key = '', value = ''] in orderedProps" :key="key" class="property-line">
+				{{ key }}: {{ value }}
```
I like it! thanks

latin-panda commented:
@garethbowen @matthew-white Thanks for the feedback! I've addressed all of the issues. 

We decided to postpone `stroke`, `marker-color`, etc for a future PR and deliver the current features ([see conversation](https://getodk.slack.com/archives/C051L6AKG4F/p1759255426244649?thread_ts=1759174897.384729&cid=C051L6AKG4F)). Tomorrow, I plan to write front-end tests and perform a thorough manual check. 

@srujner, this will be available to you soon unless significant changes are requested.

latin-panda commented on packages/web-forms/src/components/common/map/MapBlock.vue:
```diff
@@ -0,0 +1,283 @@
+					<IconSVG name="mdiArrowExpandAll" size="sm" />
+				</button>
+				<button @click="centerFeatureLocation">
+					<IconSVG name="mdiFullscreen" />
```
Yes! I've added the disabled state for that button.

> Is this actually a useful feature? What's the user story here? I've saved something then moved around and now I want to zoom back to it?
> If it IS useful, then is there a better icon for it? Initially I thought it was a full screen icon (but that's the one above).

Yes, that's how someone would use it. What icon do you think might represent a saved feature? 
Tagging @alyblenkin

latin-panda replied:
>For what it’s worth, I thought it was zoom to fit all options and that it just wasn’t working right. I did not notice that it centered the selected item. Tapping on the selected item or “view details” to center the selection both seem intuitive to me and work as expected.
I use the “zoom to fit all options” frequently in Collect for form testing and demos. I don’t know how useful it is for real data collection.

😅 Maybe there has been a confusion. From the start, we conceptualized it to center the view on the saved feature (not the selected feature or fit all features):

<img width="851" height="225" alt="Screenshot 2025-10-01 at 1 38 14 PM" src="https://github.com/user-attachments/assets/08c676f9-c2ea-4015-b84c-5ce83177e392" />

I'll prepare the change to save time. I'll wait for @alyblenkin to confirm the behavior before pushing any updates.

latin-panda replied:
Fixed! 

https://github.com/user-attachments/assets/f3938979-d2c8-4898-9333-bcb6cadcc38e

garethbowen commented on packages/web-forms/src/components/common/map/MapProperties.vue:
```diff
@@ -0,0 +1,140 @@
+		</div>
+
+		<div class="map-properties-content">
+			<dl v-for="[key = '', value = ''] in orderedProps" :key="key" class="property-line">
+				<dt>{{ key }}</dt><dd>{{ value }}</dd>
+			</dl>
```
Ouch yeah, getting the border working is a pain. I'm sorry I've made this more complicated than it needs to be. Feel free to go back to your original version  :/

matthew-white commented on packages/web-forms/src/components/common/map/useMapBlock.ts:
```diff
@@ -0,0 +1,293 @@
+		currentState.value = STATES.LOADING;
+		selectFeature(undefined);
+		saveFeature(undefined);
+		featuresSource.clear();
```
I've been calling `clear()` with argument `true`. The name of that parameter is `fast`, which sounds appealing haha. The API docs describe `fast` as:

> Skip dispatching of `removefeature` events.

As far as I know, those events aren't needed, so maybe `true` would help a little.

matthew-white commented on packages/web-forms/src/components/common/map/useMapBlock.ts:
```diff
@@ -0,0 +1,293 @@
+	const currentState = shallowRef<(typeof STATES)[keyof typeof STATES]>(STATES.LOADING);
+	const errorMessage = shallowRef<{ title: string; message: string } | undefined>();
+	let mapInstance: Map | undefined;
+	const savedFeature = ref<Feature<GeometryType> | undefined>();
+	const selectedFeature = ref<Feature<GeometryType> | undefined>();
```
Could these be shallow refs? Do you ever mutate a `Feature` object after setting it on the ref?

matthew-white commented on packages/web-forms/src/components/common/map/MapBlock.vue:
```diff
@@ -0,0 +1,291 @@
+watch(
+	() => props.disabled,
+	(newValue) => mapHandler?.toggleClickBinding(!newValue),
+	{ deep: true, immediate: true }
```
I'm having a similar thought about the watcher on `props.featureCollection` above, which calls `loadGeometries()`. `initializeMap()` is called in `onMounted()`, and `initializeMap()` calls `loadGeometries()`, so I feel like that watcher doesn't need `immediate: true`. Or alternatively, maybe `immediate: true` could be kept here, but `initializeMap()` doesn't need to call `loadGeometries()`.

matthew-white replied:
Now that I've looked at `useMapBlock()`, I'm realizing that all or almost of the watchers in this PR have `{ immediate: true }`. We don't use `{ immediate: true }` by default in Central, but if that's an established Web Forms pattern, maybe these comments aren't relevant / feel free to ignore. Personally, I feel like control flow is clearer when a watcher is only called when needed, but that might just be a personal preference.

latin-panda replied:
I'm curious 🤔 Why isn't it used in Central?

matthew-white commented on packages/web-forms/src/components/common/map/useMapBlock.ts:
```diff
@@ -0,0 +1,293 @@
+			return;
+		}
+
+		const featureToSave = featuresSource.getFeatures()?.find((feature) => {
+			return feature.getProperties()?.odk_value === value;
+		}) as Feature<GeometryType>;
```
It looks like `getFeatures()` will create a new array. I think `forEachFeature()` is a way to iterate without creating an array. Maybe not a concern, it would only matter if there are many features such that the new array would be large.

matthew-white reviewed (commented):
I didn't look at everything closely, but I thought I'd leave a few comments (all small things). The comments are just my two cents — I certainly don't want to be blocking merge or anything. I'm going to be busy tomorrow and out on Thursday, so I might not be able to reply to follow-up comments until Friday.

I need to update my Central PR to match this PR in a few places, e.g., map-styles.ts. I'll do that soon.

latin-panda commented on packages/web-forms/src/components/common/map/MapBlock.vue:
```diff
@@ -0,0 +1,291 @@
+watch(
+	() => props.featureCollection,
+	(newData) => {
+		mapHandler?.loadGeometries(newData);
```
Ah yes, when it was a ref, it could have been undefined, but now that I've refactored that part, we can remove those checks.
Fixed, thanks!

alyblenkin commented on packages/web-forms/src/components/common/map/MapProperties.vue:
```diff
@@ -0,0 +1,140 @@
+			</button>
+		</div>
+
+		<div class="map-properties-content">
```
Ah sorry the figma component must have went weird somehow. It should be vertically centered.

latin-panda replied:
Fixed!
<img width="851" height="597" alt="Screenshot 2025-10-02 at 12 40 04 AM" src="https://github.com/user-attachments/assets/625f3a0c-a294-46dd-9273-6321c3cd8fc9" />

garethbowen commented on packages/xforms-engine/src/client/BaseItem.ts:
```diff
@@ -0,0 +1,7 @@
+import type { TextRange } from './TextRange.ts';
+
+export interface BaseItem {
```
I'm wondering if this should be just `Item` now. BaseItem made sense when thinking about Rank and Select extending it, but seeing as there is now hierarchy now does it still make sense?

garethbowen replied:
I agree `Item` is a bit vague but so are all the other names I can think of  :/

`BaseItem` is fine as I can't think of anything better!

latin-panda commented:
I've finished going through the [test plan](https://github.com/getodk/web-forms/issues/305#issuecomment-3206808806) again, and I've uploaded videos and screenshots of some cases to this PR's description.

@srujner, I couldn't test on iOS. Could you please include it in your testing?

latin-panda commented on packages/web-forms/src/components/common/map/useMapBlock.ts:
```diff
@@ -0,0 +1,307 @@
+			view: new View({
+				center: DEFAULT_VIEW_CENTER,
+				zoom: MIN_ZOOM,
+				// Prevent map cloning at low zoom during panning, which disrupts feature selection.
+				multiWorld: false,
+				projection: DEFAULT_VIEW_PROJECTION,
+				extent: getProjection(DEFAULT_VIEW_PROJECTION)?.getExtent(),
```
@matthew-white You will need to add these lines

@srujner found an issue ([video](https://jam.dev/c/4807b3dd-a7f9-477e-be95-ddec5c6a36d9)) where the selection of features was disrupted when panning clone the map. You can see the fix in [my test server](https://effortless-dragon-d2ff5f.netlify.app/).
